### PR TITLE
Add Rust concept notes to Burn Book

### LINKED ISSUE #1155: Burn Book: Introduction to Rust
Add an introduction to rust section
For people who don't know Rust, following the guide should be enough to get running with Burn
Refer to the most relevant sections of the Rust book

### PR DESCRIPTION
**Pull request by laggui**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Closes #1155

### Changes

For new Rustaceans, following the guide should be enough to get started with Burn. In the previous version of the book, we simply mentioned that people should "read the first chapters of the Rust Book" first.

While it is still recommended, this PR introduces **🦀 Rust Note** indicators throughout the sections 3 (Getting Started) and 4 (Basic Workflow, aka "The Guide").

The notes are collapsible sections which provide explanations for the basic Rust concepts involved in the current context.

<details>
<summary><strong>🦀 An Example Note</strong></summary>

This note should briefly explain one concept (or in some cases, a few related concepts) particular to Rust and perhaps detail how it is applied in the current section.

External references are usually provided, pointing to the Rust Book as the main source.
</details><br>

Let me know what you think about the context notes as collapsible sections! I used the crab 🦀 emoji since the notes are supposed to be Rust specific, but it could be changed if you have any other preferences.

Also, any comments on the added notes are welcome! And if you think I missed anything let me know.

### CODE REVIEW

File: burn-book/src/basic-workflow/model.md [hunk 1]
```diff
@@ -51,14 +60,96 @@ There are two major things going on in this code sample.
...
+   [explainer below](#derive-attribute) to learn why.
+
+   For more details on traits, take a look at the
+   [associated chapter](https://doc.rust-lang.org/book/ch10-02-traits.html) in the Rust Book.
+   </details><br>
+
+   <details id="derive-attribute">
+   <summary><strong>🦀 Derive Attribute</strong></summary>
+
+   The `derive` attribute allows traits to be implemented easily by generating code that will
```
Comment (nathanielsimard) [inline burn-book/src/basic-workflow/model.md:82]: Not only for traits, you can add arbitrary code based on the provided type, but you can't modify the existing code; you can only append new elements. Using the attribute `#[my_macro]`, you can effectively update the provided code. Derive is frequently employed to recursively implement traits, where the implementation consists of the composition of all fields.

File: burn-book/src/basic-workflow/model.md [hunk 2]
```diff
@@ -87,6 +178,44 @@ impl ModelConfig {
...
+References in Rust allow us to point to a resource to access its data without owning it. The idea of
+ownership is quite core to Rust and is worth
+[reading up on](https://doc.rust-lang.org/book/ch04-00-understanding-ownership.html).
+
+In a language like C, memory management is explicit and up to the programmer, which means it is easy
+to make mistakes. In a language like Java or Python, memory management is automatic with the help of
+a garbage collector. This is very safe and straightforward, but also incurs a runtime cost.
+
+In Rust, memory management is rather unique. Aside from
+[primitive types](https://doc.rust-lang.org/rust-by-example/primitives.html), every value is _owned_
```
Comment (nathanielsimard) [inline burn-book/src/basic-workflow/model.md:205]: I think this is related to the `Copy` trait, not necessarily the primitive types (although they all implement `Copy`, making it a great example).

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Some minor comments, but I think this will be really useful for people without much Rust experience, thanks!

**Comment by laggui** [comment]
@nathanielsimard made the requested changes! Let me know if my additions are correct :)

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**